Add Openresty as an webserver option

TafkaMax opened the pull request:
I have successfully created and made Openresty use Modsecurity. Because Openresty is based on Nginx it was doable.

There are some nuances.

----

Because the official Openresty Docker Image is created differently than the Nginx one, then there is no usage of 'nginx' user. Openresty does not use 'docker-entrypoint' logic, so I copied that logic from the original Nginx repository.

If you have any more questions I am happy to answer them.

I also added a simple gitlab-ci.yml file, because that is what I use with the integrated gitlab registry, but it might be too specific to add.

Currently I only added the Alpine build option and only for linux/amd64.

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+    NGINX_VERSION \
+    LUA_JIT_VERSION 
+
+# Remove pcre2 support for openresty, until it gets update to pcre2.
```
Is this sentence correct? Did you maybe mean to write "... updated to pcre3"?

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+
+WORKDIR /sources
+
+# Download modsecurity and compile it.
```
```suggestion
# Download ModSecurity and compile it.
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+    make install; \
+    strip /usr/local/modsecurity/lib/lib*.so*
+
+# We use master Modsec version
```
```suggestion
# We use latest version of the ModSecurity nginx connector.
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+# We use master Modsec version
+# Instead of Nginx download the bundled openresty for this version of openresty base image. We can use the openresty already installed on the base image and get the compile options from there. For building the nginx module we can use the bundled nginx.
+# Specify the LUAJIT_LIB-s that are already installed in the base image.
+# use the includes from the bundled version. This is where LUA JIT VERSION is important. Because it differs from release to release (probably).
```
```suggestion
# Use the includes from the bundled version. This is where LUA_JIT_VERSION comes in to play because the version differs from release to release (probably).
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+    strip /usr/local/modsecurity/lib/lib*.so*
+
+# We use master Modsec version
+# Instead of Nginx download the bundled openresty for this version of openresty base image. We can use the openresty already installed on the base image and get the compile options from there. For building the nginx module we can use the bundled nginx.
```
Why is this necessary? Doesn't the Openresty image also include nginx?

```suggestion
# Download Openresty with bundled nginx. We can then get the compilation options for nginx from the installed Openresty and then build the ModSecurity nginx connector using the bundled nginx.
```

TafkaMax replied:
I have updated the comment.

I download the bundle mostly for the 'configure' script. I did not find this in the premade Openresty image under /usr/local/openresty. I guess they removed unneccessary build files from their image. As we need to add a new nginx module we need to download the version specific src of Openresty to do that.

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+# Also specify the LUA version used. (Comes with the latest Openresty src bundle.)
+ARG OPENRESTY_VERSION="1.21.4.2" \
+    NGINX_VERSION="1.21.4" \
+    LUA_JIT_VERSION="2.1-20230410"
```
Declare the ModSecurity version at the top so you can access it in all stages.

```suggestion
    LUA_JIT_VERSION="2.1-20230410"
ARG MODSEC_VERSION=3.0.9
```

TafkaMax replied:
Should be OK now. Added MODSEC_VERSION at the start of the dockerfile and removed from other places.

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat as build
+
+ARG MODSEC_VERSION=3.0.9 \
```
Use value of top-level ARG:

```suggestion
ARG MODSEC_VERSION
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+
+FROM openresty/openresty:${OPENRESTY_VERSION}-alpine-fat
+
+ARG MODSEC_VERSION=3.0.9
```
Use value from top-level ARG:
```suggestion
ARG MODSEC_VERSION
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+
+ARG MODSEC_VERSION=3.0.9
+
+LABEL maintainer="Felipe Zipitria <felipe.zipitria@owasp.org>"
```
If you agree to maintain this image, then please change this label.

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+COPY --from=build /etc/modsecurity.d/modsecurity.conf /etc/modsecurity.d/modsecurity.conf
+COPY --from=crs_release /opt/owasp-crs /opt/owasp-crs
+# We use the templating mechanism from the nginx image here,
+# as set up by owasp/modsecurity-docker
```
```suggestion
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+COPY --from=build /etc/modsecurity.d/unicode.mapping /etc/modsecurity.d/unicode.mapping
+COPY --from=build /etc/modsecurity.d/modsecurity.conf /etc/modsecurity.d/modsecurity.conf
+COPY --from=crs_release /opt/owasp-crs /opt/owasp-crs
+# We use the templating mechanism from the nginx image here,
```
We no longer use owasp/modsecurity-docker. I'll create a PR to remove that from out other files.
```suggestion
# We use the templating mechanism from the nginx image here.
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+COPY --from=crs_release /opt/owasp-crs /opt/owasp-crs
+# We use the templating mechanism from the nginx image here,
+# as set up by owasp/modsecurity-docker
+# Everything from templates except the nginx confiugration is the same for openresty.
```
```suggestion
# Everything from templates, except the nginx configuration, is the same for Openresty.
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+COPY nginx/templates/modsecurity.d /usr/local/openresty/nginx/templates/modsecurity.d
+# Copy the nginx configuration separately
+COPY openresty/templates/nginx.conf.template /usr/local/openresty/nginx/templates
+# 
```
```suggestion
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+COPY openresty/docker-entrypoint.d/*.sh /docker-entrypoint.d/
+COPY src/opt/modsecurity/activate-plugins.sh /docker-entrypoint.d/94-activate-plugins.sh
+COPY src/opt/modsecurity/activate-rules.sh /docker-entrypoint.d/95-activate-rules.sh
+COPY openresty/docker-entrypoint.sh /
```
Please move all the Openresty stuff to a separate block after the other `COPY` statements. That will make it easier to maintain and merge changes.

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+        sed \
+        yajl; \
+    # Install lua modules here
+    # Possible TODO make it a variable if you want to build this Dockerfile yourself.
```
```suggestion
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,253 @@
+
+HEALTHCHECK CMD /usr/local/bin/healthcheck
+
+# Openresty does not use NGINX dockerfile. 
```
```suggestion
# The Openresty image is not based on the nginx image.
```

theseion commented on .gitlab-ci.yml:
I'd rather not have this file in our repository. It's also pretty straight forward to adapt our github workflow to a gitlab style workflow.

theseion commented:
Thanks @TafkaMax. I have some comments but as it looks now, we could probably also build the Openresty image. We would update CRS and ModSecurity versions but would rely on you to update Openresty specific dependencies.

fzipi commented:
@TafkaMax Can you address Max's comments?

TafkaMax commented:
> Welcome back @TafkaMax ! Can you start addressing comments here, so we can merge this one?

Will do. I will plan do address the comments within a week.

fzipi commented:
Can you resolve the conversation for the things addressed already? Thanks! Looking better :)

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,254 @@
+# Current latest Openresty version is here.
+# Also specify the nginx core used. (Openresty, adds their own version numbers after the nginx one, so we can't use the same number for both Openresty and nginx)
+# Also specify the LUA version used. (Comes with the latest Openresty src bundle.)
+ARG OPENRESTY_VERSION="1.21.4.3" \
```
```suggestion
ARG OPENRESTY_VERSION="1.25.3.1" \
```

TafkaMax replied:
Will change. Finally openresty has been updated to a newer version of nginx.

theseion replied:
Sorry, I didn't mean to put additional work on your plate, I simply wanted the versions to be consistent across all images.

TafkaMax replied:
No worries

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,254 @@
+# Also specify the nginx core used. (Openresty, adds their own version numbers after the nginx one, so we can't use the same number for both Openresty and nginx)
+# Also specify the LUA version used. (Comes with the latest Openresty src bundle.)
+ARG OPENRESTY_VERSION="1.21.4.3" \
+    NGINX_VERSION="1.21.4" \
```
```suggestion
    NGINX_VERSION="1.25.3" \
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,254 @@
+# Also specify the LUA version used. (Comes with the latest Openresty src bundle.)
+ARG OPENRESTY_VERSION="1.21.4.3" \
+    NGINX_VERSION="1.21.4" \
+    LUA_JIT_VERSION="2.1-20230410" \
```
```suggestion
    LUA_JIT_VERSION="2.1-20231117" \
```

TafkaMax replied:
Removed LUA_JIT_VERSION as this variable is not needed anymore

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,254 @@
+    NGINX_VERSION \
+    LUA_JIT_VERSION 
+
+# Remove pcre2 compliation of ModSecurity for openresty, until openresty gets updated to support pcre2. Relevant github issue: https://github.com/openresty/lua-nginx-module/issues/1984
```
```suggestion
# Remove pcre2 compilation of ModSecurity for openresty, until openresty gets updated to support pcre2. Relevant github issue: https://github.com/openresty/lua-nginx-module/issues/1984
```

TafkaMax replied:
Because nginx was updated to 1.25 for openresty I will revisit the issue. The reason for openresty using 1.21 of nginx was because that was the last version to support pcre. The change to nginx 1.25 means that they have overcome that issue and I will need to change the compilation to use pcre2.

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,254 @@
+
+ARG MODSEC_VERSION
+
+LABEL maintainer="Taavi Ansper <taaviansperr@gmail.com>"
```
Is this e-mail address correct? It appears to contain one more `r` than necessary.

TafkaMax replied:
Yes the email is correct.

theseion commented on openresty/templates/.gitkeep:
The directory isn't empty, so you can delete this file. Or is there a reason for you to keep it?

TafkaMax replied:
This can be deleted. I will do it.

theseion commented:
Sorry for the long wait @TafkaMax. I think you done a good job and if you are taking responsibility of the image then I don't see why we shouldn't offer it as part of the default build.

TafkaMax commented:
Updated the image to version 1.25.3.

Now openresty can use pcre2 again, which is nice.

Tested in my environment and build successfully and works aswell.

theseion commented on README.md:
```diff
@@ -28,11 +28,15 @@ We also build [alpine linux](https://www.alpinelinux.org/) variants of the base
 * `3-apache-alpine-YYYYMMDDHHMM`, `3.3-apache-alpine-YYYYMMDDHHMM`, `3.3.5-apache-alpine-YYYYMMDDHHMM`, `apache-alpine` ([master/apache/Dockerfile-alpine](https://github.com/coreruleset/modsecurity-crs-docker/blob/master/apache/Dockerfile-alpine)) – *last stable ModSecurity v2 on Apache 2.4.58 official alpine stable base image, and latest stable Core Rule Set 3.3.5*
 
-⚠️ We changed tags to [support production usage](https://github.com/coreruleset/modsecurity-crs-docker/issues/67). Now, if you want to use the "rolling version", use the tag `owasp/modsecurity-crs:nginx-alpine` or `owasp/modsecurity-crs:apache-alpine`. If you need a stable long term image, use the one with the full CRS version, in addition to the build date in `YYYYMMDDHHMM` format, example `owas ...
+⚠️ We changed tags to [support production usage](https://github.com/coreruleset/modsecurity-crs-docker/issues/67). Now, if you want to use the "rolling version", use the tag `owasp/modsecurity-crs:nginx-alpine` or `owasp/modsecurity-crs:apache-alpine` or `owasp/modsecurity-crs:openresty-alpine-fat`. If you need a stable long term image, use the one with the full CRS version, in addition to the bu ...
```
```suggestion
⚠️ We changed tags to [support production usage](https://github.com/coreruleset/modsecurity-crs-docker/issues/67). Now, if you want to use the "rolling version", use the tag `owasp/modsecurity-crs:nginx-alpine` or `owasp/modsecurity-crs:apache-alpine` or `owasp/modsecurity-crs:openresty-alpine-fat`. If you need a stable long term image, use the one with the full CRS version, in addition to the build date in `YYYYMMDDHHMM` format (e.g., `owasp/modsecurity-crs:3.3.5-nginx-alpine-202209141209`).
```

theseion commented on README.md:
```diff
@@ -28,11 +28,15 @@ We also build [alpine linux](https://www.alpinelinux.org/) variants of the base
+
+### Notes regarding Openresty version of this image.
+
+The Openresty version of modsecurity-crs-docker currently only builds an image based on **Alpine Linux**. The Dockerfile for Openresty is located in the [docker-openresty github page](https://github.com/openresty/docker-openresty/blob/master/alpine/Dockerfile.fat).
```
```suggestion
We currently only provide a version of the Openresty image based on **Alpine Linux**. The Dockerfile for Openresty resides in the [docker-openresty repository](https://github.com/openresty/docker-openresty/blob/master/alpine/Dockerfile.fat).
```

theseion commented on README.md:
```diff
@@ -28,11 +28,15 @@ We also build [alpine linux](https://www.alpinelinux.org/) variants of the base
 ## Supported architectures
 
-We added the [docker buildx](https://github.com/docker/buildx) support to our docker builds so additional architectures are supported now. As we create our containers based on the official apache and nginx ones, we can only support the architectures they support.
+We added the [docker buildx](https://github.com/docker/buildx) support to our docker builds so additional architectures are supported now. As we create our containers based on the official apache, nginx and openresty ones, we can only support the architectures they support.
```
```suggestion
We added the [docker buildx](https://github.com/docker/buildx) support to our docker builds so additional architectures are supported now. As we create our containers based on the official Apache httpd, nginx and Openresty ones, we can only support the architectures they support.
```

theseion commented on README.md:
```diff
@@ -218,6 +229,10 @@ Note: Apache access and metric logs can be disabled by exporting the `nologging=
 
+### Openresty ENV Variables
+
+Openresty uses the same ENV variables as the nginx version.
```
```suggestion
Openresty uses the same environment variables as the nginx version.
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,252 @@
+    # Change this from normal nginx setup. Do not add /usr/lib or /lib
+    LD_LIBRARY_PATH=/usr/local/lib:/usr/local/openresty \
+    NGINX_ENVSUBST_OUTPUT_DIR=/usr/local/openresty/nginx/conf \
+    # overridden variables
```
Removed just recently.
```suggestion
```

theseion commented on openresty/Dockerfile-alpine:
```diff
@@ -0,0 +1,252 @@
+    LD_LIBRARY_PATH=/usr/local/lib:/usr/local/openresty \
+    NGINX_ENVSUBST_OUTPUT_DIR=/usr/local/openresty/nginx/conf \
+    # overridden variables
+    USER=root \
```
Removed just recently.
```suggestion
```

theseion commented:
Awesome, thank you!
